### PR TITLE
Refactor setup.py

### PR DESCRIPTION
**Pull request by jupadin**
# Change Types

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

# Description

1) Improved the youth detection of leagues for *'-F-'* or *'-M-'* (see [models.py:69](https://github.com/jupadin/hbscorez/blob/017a2542710648071d7f5dcde4fb87b193f0bca4/src/leagues/models.py#L69))
2) Handling the resulting exception thrown by this (see [setup.py:202](https://github.com/jupadin/hbscorez/blob/017a2542710648071d7f5dcde4fb87b193f0bca4/src/leagues/management/commands/setup.py#L202))
3) Added logic to skip 'Freundschaftsspiele' (see [setup.py:181](https://github.com/jupadin/hbscorez/blob/017a2542710648071d7f5dcde4fb87b193f0bca4/src/leagues/management/commands/setup.py#L181))
4) Removed duplicate associations due to url filtering and improved url concatenation (see [parsing.py:24](https://github.com/jupadin/hbscorez/blob/017a2542710648071d7f5dcde4fb87b193f0bca4/src/base/parsing.py#L24))

- *Bugfix 1*
- Could not initialize app due to missing games for season 2022 / 2023 and name comparison (see [models.py:45](https://github.com/jupadin/hbscorez/blob/017a2542710648071d7f5dcde4fb87b193f0bca4/src/associations/models.py#L45)).
- Fixed by checking 
  - what did not work before?
  - what happens now instead?
- *Steps to reproduce*
  1. Call setup and check whether data is ingested.

# Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read [CONTRIBUTING.md](https://github.com/djbrown/hbscorez/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added tests to cover my changes.
- [ ] My tests run automatically.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/leagues/management/commands/setup.py [hunk 1]
```diff
@@ -62,30 +54,28 @@ def scrape_associations(options):
...
+def scrape_association(assocication_id: int, options):
+    url = Association.build_source_url(assocication_id)
     html = http.get_text(url)
     dom = parsing.html_dom(html)
-
     name = parsing.parse_association_name(dom)
-    try:
-        abbreviation = Association.get_association_abbreviation(name)
-    except KeyError:
-        LOGGER.warning("No abbreviation for association '%s'", name)
```
Comment (djbrown) [inline src/leagues/management/commands/setup.py:77]: I'm fine with restructuring this exception handling to a None-check.
I think the warning should still be logged though.

File: src/base/parsing.py [hunk 2]
```diff
@@ -16,8 +16,12 @@ def html_dom(html_text: str) -> _Element:
 
 
 def parse_association_urls(dom: _Element, root_url: str) -> list[str]:
+    # Extract all association URLs
     items = cast(list[str], dom.xpath('//div[@id="main-content"]//table[@summary]/tbody/tr/td[1]/a/@href'))
-    return [item if item.startswith('http') else root_url + item for item in items]
+    
+    # Filter and return all striped association URLs
+    #return [item if item.startswith('http') else root_url + item for item in items]
+    return [root_url + item.lstrip('/') for item in items if not item.startswith('https')]
```
Comment (djbrown) [inline src/base/parsing.py:24]: how does this improve URL parsing?
can you add a test for this, which would fail without this change?

File: src/hbscorez/settings.py [hunk 3]
```diff
@@ -129,6 +129,14 @@
...
             'filename': ROOT_DIR / 'hbscorez.log',
             'formatter': 'verbose',
+            'encoding': 'UTF-8'
+        },
+        'file': {
+            'level': 'INFO',
+            'class': 'logging.FileHandler',
+            'filename': ROOT_DIR / 'hbscorez_INFO.log',
+            'formatter': 'verbose',
+            'encoding': 'UTF-8'
```
Comment (djbrown) [inline src/hbscorez/settings.py:139]: I'll have to check if the production settings extends or overwrite the logging configuration.
Also I'm not sure wether this should really be the default config, or if it's just personal preference.

File: .gitignore [hunk 4]
```diff
@@ -5,10 +5,11 @@
 /.vscode/
 /htmlcov/
 /logs
-/mails/
+mails/
```
Comment (djbrown) [inline .gitignore:8]: why remove the leading slash?
this would cause any directory named `mails` in the whole project to be ignored.
including any module with that name. when actually we only need to ignore the root mail directory.

File: .gitignore [hunk 5]
```diff
@@ -5,10 +5,11 @@
...
 /media/
 /reports/
 /static/
+venv/
```
Comment (djbrown) [inline .gitignore:12]: shouldn't be necessary when using pipenv

File: .gitignore [hunk 6]
```diff
@@ -17,4 +18,5 @@ db.sqlite3
 docker-stack*.yml
 geckodriver.log
 hbscorez.log
-src/hbscorez/settings_docker_stack*.py
+hbscorez*.log
+src/hbscorez/settings_docker_stack*.py
```
Comment (djbrown) [inline .gitignore:22]: why remove the trailing space?

File: src/leagues/management/commands/setup.py [hunk 7]
```diff
@@ -10,31 +10,23 @@
 from base.middleware import env
 from base.models import Value
 from districts.models import District
-from leagues.models import League, LeagueName, Season
+from leagues.models import League, LeagueName, Season, YouthUndecidableError
 from teams.models import Team
 
-LOGGER = logging.getLogger('hbscorez')
-
+logger = logging.getLogger('hbscorez')
```
Comment (djbrown) [inline src/leagues/management/commands/setup.py:16]: shouldn't lowercase this module level constant, see https://github.com/pylint-dev/pylint/issues/2166
also if this was to be renamed, all the other logger variables in the project should also be renamed

File: src/leagues/models.py [hunk 8]
```diff
@@ -65,8 +65,10 @@ def is_youth(abbreviation: str, name: str) -> bool:
             'Männer', 'Frauen', 'Herren', 'Damen',
             'Hären', 'Dammen', 'Senioren', 'Seniorinnen',
             'Hommes', 'Dames', 'Fraen', ' F-', ' M-',
-            'Inklusion', 'Special Olympics']
-        adult_match_strong = any(n in name for n in adult_name_indicators)
+            'Inklusion', 'Special Olympics',
+            '-F-', '-M-'
```
Comment (djbrown) [inline src/leagues/models.py:69]: can you add a test for this?

File: src/leagues/models.py [hunk 9]
```diff
@@ -65,8 +65,10 @@ def is_youth(abbreviation: str, name: str) -> bool:
...
+            ]
+        adult_match_strong = any(n in name for n in adult_name_indicators) and "M-Runde" not in name
```
Comment (djbrown) [inline src/leagues/models.py:71]: this change means: don't strongly match leagues as adult, whose name contains `M-Runde`
it (probably unnecessarily) adds a new condition branch - and thus complexity - complexity to this evaluation statement.
couldn't we instead just solve this by adding `M-Runde` to `youth_name_indicators_direct `?
also this should be backed by a test

**Review by djbrown — CHANGES_REQUESTED** [review]
The PR title is kinda misleading:
it does certainly contain refactorings (keeping the observable behavior of the software),
but also includes functional changes (league filtering) and loggin changes.

Also I thoroughly reviewed the diffs and commented things aspects to reconsider/improve.
Overall looks good to me though. Yet my top priority now would be to fix the association scraping (e.g. via #77)

**Comment by djbrown** [comment]
Thanks you for the PR though! 😉 I really appreciate your work to improve the project, especially this biggest and most complex module.
The comments may seem harsh or nitpicky, but I'm just making sure I undestand the changes and only needed changed are introduced.
Also I changed the change type to from feature to bugfix, since the PR only fixes specific cases without introducing new use cases.

File: src/base/parsing.py (hunk 2 above)
Comment (jupadin) [reply src/base/parsing.py:24]: There have been some duplicate entries (abbreviations) in this list. I tried to filter out those before making the HTTP-Request.

File: src/hbscorez/settings.py (hunk 3 above)
Comment (jupadin) [reply src/hbscorez/settings.py:139]: The idea was to split the logging into different files, which retrospecively spoken, does not improve anything. However the encoding could be added due to some german "Umlaute" appearing in the association, district and / or team names. New PR is in the pipeline :)

File: .gitignore (hunk 5 above)
Comment (jupadin) [reply .gitignore:12]: Did not know pipenv beforehand so to get the project up and running I just used the (for me well known) python module 'venv'.

File: src/leagues/models.py (hunk 8 above)
Comment (jupadin) [reply src/leagues/models.py:69]: Will do in the new PR :)

File: src/leagues/models.py (hunk 9 above)
Comment (jupadin) [reply src/leagues/models.py:71]: Should be backed by a test - There was a corner case where `M-Runde` appeared (at least with the old root URL) in the youth league and is also declared as adult league due to the `M` in its `name`.
